File: beam-android/core/ffi/src/main/kotlin/dev/beam/android/core/ffi/repository/BeamRepositories.kt
package dev.beam.android.core.ffi.repository

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import uniffi.beam_client_core.AdminEvent
import uniffi.beam_client_core.AdminLogEntry
import uniffi.beam_client_core.AdminStatus
import uniffi.beam_client_core.AdminUserPage
import uniffi.beam_client_core.BeamClient
import uniffi.beam_client_core.BrowseQuery
import uniffi.beam_client_core.ContinueWatchingEntry
import uniffi.beam_client_core.DeviceProfile
import uniffi.beam_client_core.DeviceSession
import uniffi.beam_client_core.EpisodeSummary
import uniffi.beam_client_core.HistoryPage
import uniffi.beam_client_core.LibraryFileSummary
import uniffi.beam_client_core.LibrarySummary
import uniffi.beam_client_core.MediaDetail
import uniffi.beam_client_core.MediaPage
import uniffi.beam_client_core.MediaSourceView
import uniffi.beam_client_core.PlaybackHttpConfig
import uniffi.beam_client_core.ProgressOutcome
import uniffi.beam_client_core.QualityPolicy
import uniffi.beam_client_core.ServerHealth
import uniffi.beam_client_core.ServerSummary
import uniffi.beam_client_core.SessionState
import uniffi.beam_client_core.SourceSelection
import uniffi.beam_client_core.UserSummary
import javax.inject.Inject
import javax.inject.Singleton

/**
 * The server registry, over the core.
 *
 * The core holds the registry as plain state rather than as a stream, so this
 * re-emits the whole list after any mutation. A media app has a handful of
 * servers at most, which makes a diffing protocol across the FFI boundary
 * more machinery than the problem deserves.
 */
@Singleton
internal class BeamServerRepository @Inject constructor(
    private val client: BeamClient,
) : ServerRepository {

    private val changes = MutableSharedFlow<List<ServerSummary>>(replay = 1)

    override val servers: Flow<List<ServerSummary>> = changes.asSharedFlow()

    override suspend fun restore(): List<ServerSummary> =
        client.restore().also { changes.emit(it) }

    override suspend fun addServer(baseUrl: String, displayName: String?): ServerSummary =
        client.addServer(baseUrl, displayName).also { republish() }

    override suspend fun selectServer(serverId: String) {
        client.selectServer(serverId)
        republish()
    }

    override suspend fun removeServer(serverId: String) {
        client.removeServer(serverId)
        republish()
    }

    override suspend fun loginUrl(serverId: String): String = client.loginUrl(serverId)

    override suspend fun completeLogin(serverId: String, sessionCookie: String): UserSummary =
        client.completeLogin(serverId, sessionCookie).also { republish() }

    override suspend fun logout(serverId: String) {
        client.logout(serverId)
        republish()
    }

    override suspend fun sessionState(serverId: String): SessionState =
        client.sessionState(serverId)

    override suspend fun activeServer(): ServerSummary? =
        client.listServers().firstOrNull { it.isActive }

    private suspend fun republish() {
        changes.emit(client.listServers())
    }
}

@Singleton
internal class BeamCatalogRepository @Inject constructor(
    private val client: BeamClient,
) : CatalogRepository {
    override suspend fun browse(query: BrowseQuery): MediaPage = client.browseMedia(query)
    override suspend fun detail(mediaId: String): MediaDetail = client.mediaDetail(mediaId)
    override suspend fun genres(): List<String> = client.genres()
    override suspend fun libraries(): List<LibrarySummary> = client.libraries()
    override suspend fun library(libraryId: String): LibrarySummary = client.library(libraryId)
    override suspend fun libraryFiles(libraryId: String): List<LibraryFileSummary> =
        client.libraryFiles(libraryId)

    override suspend fun upNext(showId: String, currentEpisodeId: String): EpisodeSummary? =
        client.upNextInShow(showId, currentEpisodeId)
}

@Singleton
internal class BeamPlaybackRepository @Inject constructor(
    private val client: BeamClient,
) : PlaybackRepository {
    override suspend fun setDeviceProfile(profile: DeviceProfile) {
        client.setDeviceProfile(profile)
    }

    override suspend fun sources(mediaId: String): List<MediaSourceView> =
        client.mediaSources(mediaId)

    override suspend fun selectSource(mediaId: String, policy: QualityPolicy): SourceSelection =
        client.selectPlaybackSource(mediaId, policy)

    override suspend fun playbackConfig(fileId: String): PlaybackHttpConfig =
        client.playbackConfig(fileId)

    override suspend fun continueWatching(limit: UInt?): List<ContinueWatchingEntry> =
        client.continueWatching(limit)

    override suspend fun history(limit: UInt?, offset: UInt?): HistoryPage =
        client.history(limit, offset)

    override suspend fun reportProgress(
        fileId: String,
        positionSecs: Double,
        durationSecs: Double?,
        force: Boolean,
    ): ProgressOutcome = client.reportProgress(fileId, positionSecs, durationSecs, force)

    override suspend fun flushProgress(): UInt = client.flushProgress()

    override suspend fun pendingProgressCount(): UInt = client.pendingProgressCount()
}

@Singleton
internal class BeamSessionRepository @Inject constructor(
    private val client: BeamClient,
) : SessionRepository {
    override suspend fun sessions(): List<DeviceSession> = client.sessions()
    override suspend fun revoke(sessionId: String) = client.revokeSession(sessionId)
    override suspend fun logoutEverywhere() = client.logoutEverywhere()
}

@Singleton
internal class BeamAdminRepository @Inject constructor(
    private val client: BeamClient,
) : AdminRepository {
    override suspend fun status(): AdminStatus = client.adminStatus()
    override suspend fun health(): ServerHealth = client.health()
    override suspend fun users(limit: UInt?, offset: UInt?): AdminUserPage =
        client.adminUsers(limit, offset)

    override suspend fun setUserDisabled(userId: String, disabled: Boolean) =
        client.setUserDisabled(userId, disabled)

    override suspend fun logs(limit: UInt?, offset: UInt?): List<AdminLogEntry> =
        client.adminLogs(limit, offset)

    override suspend fun logCount(): ULong = client.adminLogCount()
    override suspend fun events(limit: UInt?): List<AdminEvent> = client.adminEvents(limit)

    override suspend fun createLibrary(name: String, rootPath: String): LibrarySummary =
        client.createLibrary(name, rootPath)

    override suspend fun deleteLibrary(libraryId: String) = client.deleteLibrary(libraryId)
    override suspend fun scanLibrary(libraryId: String): UInt = client.scanLibrary(libraryId)
    override suspend fun refreshMetadata(mediaId: String) = client.refreshMediaMetadata(mediaId)
}

File: beam-apple/Packages/BeamKit/Sources/BeamCore/BeamRepositories.swift
import BeamFFI
import BeamModel
import Foundation

/// Every repository, over one `BeamClient`.
///
/// One type rather than six, because the six protocols are six views of one
/// object and splitting the adapter would mean six copies of the same
/// `mapping` helper. The protocols stay separate so a screen declares only the
/// surface it uses and a fake only has to implement that much.
public struct BeamRepositories: ServerRepository, CatalogRepository, PlaybackRepository,
    SessionRepository, AdminRepository
{
    private let client: BeamClient

    /// Wrap a client.
    public init(client: BeamClient) {
        self.client = client
    }

    // MARK: - ServerRepository

    public func restore() async throws -> [ServerSummary] {
        try await mapping { try await client.restore() }
    }

    public func listServers() throws -> [ServerSummary] {
        try mappingSync { try client.listServers() }
    }

    public func addServer(baseURL: String, displayName: String?) async throws -> ServerSummary {
        try await mapping { try await client.addServer(baseUrl: baseURL, displayName: displayName) }
    }

    public func selectServer(id serverId: String) async throws {
        try await mapping { try await client.selectServer(serverId: serverId) }
    }

    public func removeServer(id serverId: String) async throws {
        try await mapping { try await client.removeServer(serverId: serverId) }
    }

    public func loginURL(serverId: String) throws -> String {
        try mappingSync { try client.loginUrl(serverId: serverId) }
    }

    public func completeLogin(serverId: String, sessionCookie: String) async throws -> UserSummary {
        try await mapping {
            try await client.completeLogin(serverId: serverId, sessionCookie: sessionCookie)
        }
    }

    public func logout(serverId: String) async throws {
        try await mapping { try await client.logout(serverId: serverId) }
    }

    public func sessionState(serverId: String) throws -> SessionState {
        try mappingSync { try client.sessionState(serverId: serverId) }
    }

    public func trustCertificate(serverId: String, fingerprint: String) async throws {
        try await mapping {
            try await client.trustCertificate(serverId: serverId, fingerprint: fingerprint)
        }
    }

    public func forgetCertificates(serverId: String) async throws {
        try await mapping { try await client.forgetCertificates(serverId: serverId) }
    }

    public func trustedCertificates(serverId: String) throws -> [String] {
        try mappingSync { try client.trustedCertificates(serverId: serverId) }
    }

    // MARK: - CatalogRepository

    public func browse(query: BrowseQuery) async throws -> MediaPage {
        try await mapping { try await client.browseMedia(query: query) }
    }

    public func detail(mediaId: String) async throws -> MediaDetail {
        try await mapping { try await client.mediaDetail(mediaId: mediaId) }
    }

    public func genres() async throws -> [String] {
        try await mapping { try await client.genres() }
    }

    public func upNext(showId: String, currentEpisodeId: String) async throws -> EpisodeSummary? {
        try await mapping {
            try await client.upNextInShow(showId: showId, currentEpisodeId: currentEpisodeId)
        }
    }

    public func libraries() async throws -> [LibrarySummary] {
        try await mapping { try await client.libraries() }
    }

    public func library(id libraryId: String) async throws -> LibrarySummary {
        try await mapping { try await client.library(libraryId: libraryId) }
    }

    public func libraryFiles(libraryId: String) async throws -> [LibraryFileSummary] {
        try await mapping { try await client.libraryFiles(libraryId: libraryId) }
    }

    // MARK: - PlaybackRepository

    public func setDeviceProfile(_ profile: DeviceProfile) {
        client.setDeviceProfile(profile: profile)
    }

    public func sources(mediaId: String) async throws -> [MediaSourceView] {
        try await mapping { try await client.mediaSources(mediaId: mediaId) }
    }

    public func selectSource(mediaId: String, policy: QualityPolicy) async throws -> SourceSelection
    {
        try await mapping {
            try await client.selectPlaybackSource(mediaId: mediaId, policy: policy)
        }
    }

    public func playbackConfig(fileId: String) throws -> PlaybackHttpConfig {
        try mappingSync { try client.playbackConfig(fileId: fileId) }
    }

    public func serverHTTPConfig() throws -> ServerHttpConfig {
        try mappingSync { try client.serverHttpConfig() }
    }

    public func reportProgress(
        fileId: String,
        positionSeconds: Double,
        durationSeconds: Double?,
        force: Bool
    ) async throws -> ProgressOutcome {
        try await mapping {
            try await client.reportProgress(
                fileId: fileId,
                positionSecs: positionSeconds,
                durationSecs: durationSeconds,
                force: force
            )
        }
    }

    public func flushProgress() async throws -> UInt32 {
        try await mapping { try await client.flushProgress() }
    }

    public func pendingProgressCount() async throws -> UInt32 {
        try await mapping { try await client.pendingProgressCount() }
    }

    public func continueWatching(limit: UInt32?) async throws -> [ContinueWatchingEntry] {
        try await mapping { try await client.continueWatching(limit: limit) }
    }

    public func history(limit: UInt32?, offset: UInt32?) async throws -> HistoryPage {
        try await mapping { try await client.history(limit: limit, offset: offset) }
    }

    // MARK: - SessionRepository

    public func sessions() async throws -> [DeviceSession] {
        try await mapping { try await client.sessions() }
    }

    public func revoke(sessionId: String) async throws {
        try await mapping { try await client.revokeSession(sessionId: sessionId) }
    }

    public func logoutEverywhere() async throws {
        try await mapping { try await client.logoutEverywhere() }
    }

    // MARK: - AdminRepository

    public func status() async throws -> AdminStatus {
        try await mapping { try await client.adminStatus() }
    }

    public func users(limit: UInt32?, offset: UInt32?) async throws -> AdminUserPage {
        try await mapping { try await client.adminUsers(limit: limit, offset: offset) }
    }

    public func setUserDisabled(userId: String, disabled: Bool) async throws {
        try await mapping { try await client.setUserDisabled(userId: userId, disabled: disabled) }
    }

    public func logs(limit: UInt32?, offset: UInt32?) async throws -> [AdminLogEntry] {
        try await mapping { try await client.adminLogs(limit: limit, offset: offset) }
    }

    public func logCount() async throws -> UInt64 {
        try await mapping { try await client.adminLogCount() }
    }

    public func events(limit: UInt32?) async throws -> [AdminEvent] {
        try await mapping { try await client.adminEvents(limit: limit) }
    }

    public func createLibrary(name: String, rootPath: String) async throws -> LibrarySummary {
        try await mapping { try await client.createLibrary(name: name, rootPath: rootPath) }
    }

    public func deleteLibrary(id libraryId: String) async throws {
        try await mapping { try await client.deleteLibrary(libraryId: libraryId) }
    }

    public func scanLibrary(id libraryId: String) async throws -> UInt32 {
        try await mapping { try await client.scanLibrary(libraryId: libraryId) }
    }

    public func refreshMetadata(mediaId: String) async throws {
        try await mapping { try await client.refreshMediaMetadata(mediaId: mediaId) }
    }

    public func health() async throws -> ServerHealth {
        try await mapping { try await client.health() }
    }

    // MARK: - Error mapping

    // Every call goes through one of these two, so a screen can never receive
    // a raw `BeamError`. Doing it at each call site instead would mean one
    // forgotten `catch` is a crash-shaped error message in front of a person.

    private func mapping<T>(_ body: () async throws -> T) async throws -> T {
        do {
            return try await body()
        } catch {
            throw BeamFailure.from(error)
        }
    }

    private func mappingSync<T>(_ body: () throws -> T) throws -> T {
        do {
            return try body()
        } catch {
            throw BeamFailure.from(error)
        }
    }
}
